exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillFileSystem.java

```diff
@@ -83,28 +87,63 @@
   private final OperatorStats operatorStats;
   private final CompressionCodecFactory codecFactory;
 
+  private boolean initialized = false;
+
   public DrillFileSystem(Configuration fsConf) throws IOException {
     this(fsConf, null);
   }
 
   public DrillFileSystem(Configuration fsConf, OperatorStats operatorStats) throws IOException {
+    Preconditions.checkNotNull(fsConf);
+
+    // Configuration objects are mutable, and the underlying FileSystem object may directly use a passed in Configuration.
+    // In order to avoid scenarios where a Configuration can change after a DrillFileSystem is created, we make a copy
+    // of the Configuration.
+    fsConf = new Configuration(fsConf);
     this.underlyingFs = FileSystem.get(fsConf);
```

Comment:
Looks like it can return underlyingFs from cache. I think it means that someone else can get this Fs and change it. In the result Drill will rely on changed underlyingFs. Possibly we need copy of underlyingFs too.

Reply:
There is a complication with this. The readers currently don't close the FileSystem after they are done with it. If we were to create a new underlyingFs each time we make a DrillFileSystem we would have to make sure that we close it after we are done reading and writing. This would require updating the readers and writers that use DrillFileSystem.

I feel that would be a big change to do and I would like to do that incrementally as separate Jiras. What are your thoughts?

Reply:
Agree. Just leave TODO here with a note and number of Jira

Reply:
Added a TODO to the javadoc in DrillFileSystemCache.java and created https://issues.apache.org/jira/browse/DRILL-6608.